### PR TITLE
fix(schema-plugin): cannot display constraint name for ob oralce 4.2.1

### LINKED ISSUE #519: [Bug]: npe will be thrown when detail a table with unnamed constraints for oracle mode on OceanBase 4.2
### ODC version

4.2.0

### OB version

4.2.1

### What happened?

create a table as follow:
```SQL
create table out_of_line_tbl(
    id integer,
    name varchar2(64),
    PRIMARY KEY(id),
    unique(name)
);
```

double click this table to show it's detail:
截屏2023-10-16 16 57 12

### What did you expect to happen?

do not thrown npe

### How can we reproduce it (as minimally and precisely as possible)?

follow 'what happened'

### Anything else we need to know?

_No response_

### Cloud

_No response_

### LINKED ISSUE #523: [Enhancement]: take a more general approach to obtaining index information
### What would you like to be added?

We use virtual table to obtain index info which is not so good, there are two reasons:
1. virtual table is in sys schema which need extra privileges.
2. OceanBase kernel does not guarantee compatibility.

We'd better use another general way to obtain the index information.

### Why is this needed?

follow 'what would you like to be added'.

### PR DESCRIPTION
**Pull request by PeachThinking**
#### What type of PR is this?
type-bug
module-Database Object

#### What this PR does / why we need it:
ob 4.2.1 oracle new feature: system-generated constraint names will not be displayed in table ddl.
1. The `OBOracleGetDBTableByParser#listConstraints` method relies on the constraint name in the table's ddl. After ob 4.2.1, the name of each constraint cannot be obtained from the table ddl. In this case, the constraint information can only be obtained by querying the internal table as before. Therefore, time-consuming optimization cannot be done here with column constraints.
2.  Remove `OBOracleGetDBTableByParser#listIndexes` method's dependence on virtual table. Compared with the previous sql, the sql for obtaining the index in this pr can still optimize the time consumption：
image

4. fix `OBOracleGetDBTableByParser#removeIdentifiers` method may cause npe.
5. Optimize the issue of the accessor.getTableOptions() method being called repeatedly in `OBOracleTableExtension#getDetail` and `OBOracleTableExtension#getTableDDL`

#### Which issue(s) this PR fixes:

Fixes #519
Fixes #523

#### Additional documentation e.g., usage docs, etc.:

```docs

```

### CODE REVIEW

File: server/plugins/schema-plugin-ob-oracle/src/main/java/com/oceanbase/odc/plugin/schema/oboracle/parser/OBOracleGetDBTableByParser.java [hunk 1]
```diff
@@ -276,44 +282,64 @@ public List<DBTableIndex> listIndexes() {
...
+            // Get index type、global/local attribute and column names by parse index ddl.
+            try {
+                JdbcOperationsUtil.getJdbcOperations(connection).query(getIndexDDLSql, (rs) -> {
+                    String ddl = rs.getString("DDL");
+                    /**
+                     * Obtaining the DDL of the primary key through dbms_metadata.get_ddl will return the DDL of the
+                     * table, which is a kernel bug.
+                     */
+                    if (com.oceanbase.tools.dbbrowser.util.StringUtils.isNotBlank(ddl)
+                            && ddl.startsWith("CREATE TABLE")) {
```
Comment (yhilmare) [inline server/plugins/schema-plugin-ob-oracle/src/main/java/com/oceanbase/odc/plugin/schema/oboracle/parser/OBOracleGetDBTableByParser.java:296]: what's this mean? start with 'CREATE TABLE'?

File: server/plugins/schema-plugin-ob-oracle/src/main/java/com/oceanbase/odc/plugin/schema/oboracle/parser/OBOracleGetDBTableByParser.java [hunk 2]
```diff
@@ -276,44 +282,64 @@ public List<DBTableIndex> listIndexes() {
...
+                        idx.setPrimary(true);
+                        idx.setUnique(true);
+                        idx.setGlobal(true);
+                        idx.setColumnNames(
+                                primaryKey.getColumnNames().stream().map(item -> removeIdentifiers(item)).collect(
+                                        Collectors.toList()));
+                    } else {
+                        CreateIndex createIndexStmt = parseIndexDDL(ddl);
+                        idx.setGlobal(createIndexStmt.getIndexOptions().getGlobal());
+                        if (createIndexStmt.getColumns().get(0).getColumn() instanceof RelationReference
```
Comment (yhilmare) [inline server/plugins/schema-plugin-ob-oracle/src/main/java/com/oceanbase/odc/plugin/schema/oboracle/parser/OBOracleGetDBTableByParser.java:307]: Can the expr be a instance of `ColumnReference`? can u supply an example?

File: server/plugins/schema-plugin-ob-oracle/src/main/java/com/oceanbase/odc/plugin/schema/oboracle/parser/OBOracleGetDBTableByParser.java [hunk 3]
```diff
@@ -276,44 +282,64 @@ public List<DBTableIndex> listIndexes() {
...
+                            if (createIndexStmt.isUnique()) {
+                                idx.setType(DBIndexType.UNIQUE);
+                                idx.setPrimary(false);
+                                idx.setUnique(true);
+                            } else {
+                                idx.setType(DBIndexType.NORMAL);
+                                idx.setPrimary(false);
+                                idx.setUnique(false);
+                            }
+                            idx.setColumnNames(createIndexStmt.getColumns().stream()
```
Comment (yhilmare) [inline server/plugins/schema-plugin-ob-oracle/src/main/java/com/oceanbase/odc/plugin/schema/oboracle/parser/OBOracleGetDBTableByParser.java:318]: if the create index statement is as follow, how to handle it?

```SQL
create index idx_name on idx_tbl("IDX_TBL"."COL");
```

**Comment by yhilmare** [comment]
the title of this pr is not so good~~

File: server/plugins/schema-plugin-ob-oracle/src/main/java/com/oceanbase/odc/plugin/schema/oboracle/parser/OBOracleGetDBTableByParser.java (hunk 1 above)
Comment (PeachThinking) [reply server/plugins/schema-plugin-ob-oracle/src/main/java/com/oceanbase/odc/plugin/schema/oboracle/parser/OBOracleGetDBTableByParser.java:296]: In oralce 4.x, Obtaining the DDL of the primary key through dbms_metadata.get_ddl will return the DDL of the table, which is a kernel bug.

File: server/plugins/schema-plugin-ob-oracle/src/main/java/com/oceanbase/odc/plugin/schema/oboracle/parser/OBOracleGetDBTableByParser.java (hunk 3 above)
Comment (PeachThinking) [reply server/plugins/schema-plugin-ob-oracle/src/main/java/com/oceanbase/odc/plugin/schema/oboracle/parser/OBOracleGetDBTableByParser.java:318]: test ob 4.2.1 does not support this syntax：
image

File: server/plugins/schema-plugin-ob-oracle/src/main/java/com/oceanbase/odc/plugin/schema/oboracle/parser/OBOracleGetDBTableByParser.java [hunk 4]
```diff
@@ -184,19 +184,22 @@ public List<DBTableConstraint> listConstraints() {
...
                 constraint.setEnabled(
-                        ofLineConstraint.getState() == null || ofLineConstraint.getState().getEnable() == null
-                                || ofLineConstraint.getState().getEnable());
+                        outOfLineConstraint.getState() == null || outOfLineConstraint.getState().getEnable() == null
+                                || outOfLineConstraint.getState().getEnable());
                 constraint.setValidate(
-                        ofLineConstraint.getState() == null || ofLineConstraint.getState().getValidate() == null
-                                || ofLineConstraint.getState().getValidate());
-                if (ofLineConstraint instanceof OutOfLineCheckConstraint) {
+                        outOfLineConstraint.getState() == null || outOfLineConstraint.getState().getValidate() == null
```
Comment (yhilmare) [inline server/plugins/schema-plugin-ob-oracle/src/main/java/com/oceanbase/odc/plugin/schema/oboracle/parser/OBOracleGetDBTableByParser.java:195]: leak of setColumnNames when check constraints exists?

File: server/plugins/schema-plugin-ob-oracle/src/main/java/com/oceanbase/odc/plugin/schema/oboracle/parser/OBOracleGetDBTableByParser.java [hunk 5]
```diff
@@ -264,56 +268,84 @@ public List<DBTableIndex> listIndexes() {
...
+         * The ddl of the primary key can not be obtained through dbms_metadata.get_ddl(), we get columns
+         * list of primary key by parse table ddl.
          */
         if (this.constraints.size() == 0) {
             listConstraints();
         }
         List<DBTableConstraint> priConstraint = this.constraints.stream().filter(
                 constraint -> constraint.getType().equals(DBConstraintType.PRIMARY_KEY)).collect(Collectors.toList());
-        DBTableConstraint primaryKey = priConstraint.size() > 0 ? priConstraint.get(0) : null;
+        AtomicReference<String> primaryKeyName = new AtomicReference<>();
```
Comment (yhilmare) [inline server/plugins/schema-plugin-ob-oracle/src/main/java/com/oceanbase/odc/plugin/schema/oboracle/parser/OBOracleGetDBTableByParser.java:279]: why use atomic ref?

**Review by yhilmare — COMMENTED** [review]
does ob-mysql got the same problem?

File: server/plugins/schema-plugin-ob-oracle/src/main/java/com/oceanbase/odc/plugin/schema/oboracle/parser/OBOracleGetDBTableByParser.java (hunk 4 above)
Comment (PeachThinking) [reply server/plugins/schema-plugin-ob-oracle/src/main/java/com/oceanbase/odc/plugin/schema/oboracle/parser/OBOracleGetDBTableByParser.java:194]: check constraints do not need to setColumnNames ，the front end only displays `checkClause`

File: server/plugins/schema-plugin-ob-oracle/src/main/java/com/oceanbase/odc/plugin/schema/oboracle/parser/OBOracleGetDBTableByParser.java (hunk 5 above)
Comment (PeachThinking) [reply server/plugins/schema-plugin-ob-oracle/src/main/java/com/oceanbase/odc/plugin/schema/oboracle/parser/OBOracleGetDBTableByParser.java]: fixed

**Comment by yizhouxw** [comment]
> the title of this pr is not so good~`

+1, may remove some word @PeachThinking 

image

**Comment by PeachThinking** [comment]
> > the title of this pr is not so good~`
> 
> +1, may remove some word @PeachThinking
> 
> image

done～

**Pull request merged**